pkg/controller/snapshotter/snapshotter.go

```diff
@@ -273,38 +286,72 @@ func (vs *volumeSnapshotter) getSnapshotCreateFunc(snapshotName string, snapshot
 			Metadata: metav1.ObjectMeta{
 				Name: snapName,
 			},
-			Spec: crdv1.VolumeSnapshotDataSpec{
-				VolumeSnapshotRef: &v1.ObjectReference{
-					Kind: "VolumeSnapshot",
-					Name: snapshotName,
-				},
-				PersistentVolumeRef: &v1.ObjectReference{
-					Kind: "PersistentVolume",
-					Name: pvName,
-				},
-				VolumeSnapshotDataSource: *snapshotDataSource,
-			},
 			Status: crdv1.VolumeSnapshotDataStatus{
 				Conditions: []crdv1.VolumeSnapshotDataCondition{
 					readyCondition,
 				},
 			},
 		}
+
 		var result crdv1.VolumeSnapshotData
-		err = vs.restClient.Post().
-			Resource(crdv1.VolumeSnapshotDataResourcePlural).
-			Namespace(v1.NamespaceDefault).
-			Body(snapshotData).
-			Do().Into(&result)
+		for i := 0; i < createVolumeSnapshotDataRetryCount; i++ {
+			err = vs.restClient.Post().
+				Resource(crdv1.VolumeSnapshotDataResourcePlural).
+				Namespace(v1.NamespaceDefault).
+				Body(snapshotData).
+				Do().Into(&result)
+			if err != nil {
+				// Re-Try it as errors writing to the API server are common
+				glog.Infof("Error creating the VolumeSnapshotData %s: %v. Retrying...", snapshotName, err)
+				time.Sleep(createVolumeSnapshotDataInterval)
+			} else {
+				break
+			}
+		}
 
 		if err != nil {
-			//FIXME: Errors writing to the API server are common: this needs to be re-tried
 			glog.Errorf("Error creating the VolumeSnapshotData %s: %v", snapshotName, err)
+			// Don't proceed to create snapshot using the plugin due to error creating
+			// VolumeSnapshotData
+			return fmt.Errorf("Failed to create the VolumeSnapshotData %s for snapshot %s", snapName, snapshotName)
 		}
+
+		// Create a snapshot on the storage backend using the plugin
+		snapshotDataSource, err := vs.takeSnapshot(pv)
```

Comment:
I think the key here is when a VolumeSnapshot object is coming in, controller needs to determine whether a real snapshot is already taken or on-the fly. Since we update VolumeSnapshot status after the action is taken, we need a way to have some records before taking the snapshot. Just suggested above, we use metadata in VolumeSnapshot object. 
1. Get a VolumeSnapshot, check the metadata, if it is empty, first create the metadata and update this information in the API server to VolumeSnapshot, and then trigger a snapshot request.
2. If metadata is already filled, use this information to check the cloud provider to see whether a snapshot is already taken. If it is not, trigger a snapshot. Otherwise, based on the response, update VolumeSnapshot and VolumeSnapshotData.

I think it can avoid duplicate snapshot taken. The concern is the performance. It requires an API update before triggering the snapshot